File: android/core/src/main/kotlin/me/raddatz/cindy/core/Exercise.kt
package me.raddatz.cindy.core

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import me.raddatz.cindy.core.signal.RepDirection

/**
 * The movements of the "Cindy" WOD, in the order they are performed, plus the plank.
 *
 * User-facing names live in the app layer. iOS keys: `exercise.<rawValue>.plural` and
 * `exercise.<rawValue>.singular`; the unit next to a target is `unit.seconds` for holds and
 * `unit.reps` otherwise.
 */
@Serializable
enum class Exercise(val rawValue: String) {
    /** "Pull-ups" / "Pull-up". */
    @SerialName("pullUp")
    PULL_UP("pullUp"),

    /** "Push-ups" / "Push-up". */
    @SerialName("pushUp")
    PUSH_UP("pushUp"),

    /** "Squats" / "Squat". */
    @SerialName("squat")
    SQUAT("squat"),

    /** Time-based hold: the face stays at a constant distance, the clock runs. "Plank". */
    @SerialName("plank")
    PLANK("plank");

    val id: String get() = rawValue

    /** true for exercises measured in seconds instead of reps. */
    val isHold: Boolean get() = this == PLANK

    /** Default target per round: reps for movements, seconds for holds. */
    val defaultTarget: Int
        get() = when (this) {
            PULL_UP -> 5
            PUSH_UP -> 10
            SQUAT -> 15
            PLANK -> 30
        }

    /**
     * Whether the singular name matches `count` — "1 pull-up" vs "5 pull-ups" (iOS `name(for:)`).
     * German and English both get by with these two forms.
     */
    fun usesSingularName(count: Int): Boolean = count == 1

    /**
     * Expected direction of the primary (face) signal relative to the rest position.
     * Used when no calibration is available (debug mode with hard-coded thresholds).
     * Calibration measures the real direction and overrides this.
     */
    val defaultRepDirection: RepDirection
        get() = when (this) {
            PULL_UP -> RepDirection.TROUGH // hanging = face closest → largest area; chin over bar = smaller
            PUSH_UP -> RepDirection.PEAK // top = face far → small area; bottom = close → large
            SQUAT -> RepDirection.PEAK // standing = far; bottom of squat = closer
            PLANK -> RepDirection.PEAK // unused: holds use a band, not a cycle
        }

    companion object {
        fun fromRawValue(rawValue: String): Exercise? = entries.firstOrNull { it.rawValue == rawValue }
    }
}

File: ios/Cindy/Core/Exercise.swift
import Foundation

/// The three movements of the "Cindy" WOD, in the order they are performed.
enum Exercise: String, Codable, CaseIterable, Identifiable, Sendable {
    case pullUp
    case pushUp
    case squat
    /// Time-based hold on a timer the athlete starts; the camera measures nothing.
    case plank

    var id: String { rawValue }

    /// true for exercises measured in seconds instead of reps.
    var isHold: Bool { self == .plank }

    /// Holds run on a timer the athlete starts, so only rep exercises are calibrated.
    var needsCalibration: Bool { !isHold }

    /// Default target per round: reps for movements, seconds for holds.
    var defaultTarget: Int {
        switch self {
        case .pullUp: return 5
        case .pushUp: return 10
        case .squat: return 15
        case .plank: return 30
        }
    }

    /// Unit abbreviation shown next to a target.
    var unit: String { isHold ? L("unit.seconds") : L("unit.reps") }

    /// User-facing name, plural form.
    var displayName: String {
        switch self {
        case .pullUp: return L("exercise.pullUp.plural")
        case .pushUp: return L("exercise.pushUp.plural")
        case .squat: return L("exercise.squat.plural")
        case .plank: return L("exercise.plank.plural")
        }
    }

    /// Name matching `count` — "1 pull-up" vs "5 pull-ups". German and English
    /// both get by with these two forms; a language with more plural categories
    /// would need per-exercise rules in Localizable.stringsdict.
    func name(for count: Int) -> String {
        count == 1 ? singularName : displayName
    }

    /// User-facing name, singular form (a single rep during calibration).
    var singularName: String {
        switch self {
        case .pullUp: return L("exercise.pullUp.singular")
        case .pushUp: return L("exercise.pushUp.singular")
        case .squat: return L("exercise.squat.singular")
        case .plank: return L("exercise.plank.singular")
        }
    }

    /// Expected direction of the primary (face) signal relative to the rest position.
    /// Used when no calibration is available (debug mode with hard-coded thresholds).
    /// Calibration measures the real direction and overrides this.
    var defaultRepDirection: RepDirection {
        switch self {
        case .pullUp: return .trough // hanging = face closest → largest area; chin over bar = smaller
        case .pushUp: return .peak   // top = face far → small area; bottom = close → large
        case .squat: return .peak    // standing = far; bottom of squat = closer
        case .plank: return .peak    // unused: holds run on a timer
        }
    }
}
